// generic-game/dist/classes/general/status/Characteristic.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var Characteristic = function () {
  function Characteristic(_ref) {
    var name = _ref.name,
        _ref$value = _ref.value,
        value = _ref$value === undefined ? 0 : _ref$value;

    _classCallCheck(this, Characteristic);

    this._name = name;
    this._value = value;
  }

  _createClass(Characteristic, [{
    key: 'getName',
    value: function getName() {
      return this._name;
    }
  }, {
    key: 'setName',
    value: function setName(name) {
      this._name = name;
    }
  }, {
    key: 'getValue',
    value: function getValue() {
      return this._value;
    }
  }, {
    key: 'setValue',
    value: function setValue(value) {
      if (isNaN(parseFloat(value))) throw new Error('Value must be a number');
      this._value = value;
    }
  }, {
    key: 'putValue',
    value: function putValue(value) {
      if (isNaN(parseFloat(value))) throw new Error('Value must be a number');
      this.setValue(this.getValue() + value);
    }
  }]);

  return Characteristic;
}();

exports.default = Characteristic;
module.exports = exports['default'];
