// generic-game/dist/classes/general/status/Currency.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var Currency = function () {
  function Currency(_ref) {
    var name = _ref.name,
        symbol = _ref.symbol,
        _ref$value = _ref.value,
        value = _ref$value === undefined ? 0 : _ref$value;

    _classCallCheck(this, Currency);

    this.setValue(value);
    this._name = name;
    this._symbol = symbol;
  }

  _createClass(Currency, [{
    key: 'getName',
    value: function getName() {
      return this._name;
    }
  }, {
    key: 'setName',
    value: function setName(name) {
      this._name = name;
    }
  }, {
    key: 'getSymbol',
    value: function getSymbol() {
      return this._symbol;
    }
  }, {
    key: 'setSymbol',
    value: function setSymbol(symbol) {
      this._symbol = symbol;
    }
  }, {
    key: 'setValue',
    value: function setValue(value) {
      if (typeof value !== 'number') throw new Error('Currency value must be a number');
      this._value = value;
    }
  }, {
    key: 'putValue',
    value: function putValue(value) {
      this._value += value;
    }
  }, {
    key: 'getValue',
    value: function getValue() {
      return this._value;
    }
  }, {
    key: 'format',
    value: function format() {
      return this._symbol + ' ' + this._value;
    }
  }]);

  return Currency;
}();

exports.default = Currency;
module.exports = exports['default'];
