// generic-game/dist/classes/character/identity/Identity.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var Identity = function () {
  function Identity(_ref) {
    var _ref$name = _ref.name,
        name = _ref$name === undefined ? '' : _ref$name,
        _ref$gender = _ref.gender,
        gender = _ref$gender === undefined ? '' : _ref$gender,
        _ref$type = _ref.type,
        type = _ref$type === undefined ? '' : _ref$type,
        _ref$role = _ref.role,
        role = _ref$role === undefined ? '' : _ref$role;

    _classCallCheck(this, Identity);

    this._name = name;
    this._type = type;
    this._role = role;
    this._gender = gender;
  }

  _createClass(Identity, [{
    key: 'setGender',
    value: function setGender(gender) {
      this._gender = gender;
    }
  }, {
    key: 'getGender',
    value: function getGender() {
      return this._gender;
    }
  }, {
    key: 'setName',
    value: function setName(name) {
      this._name = name;
    }
  }, {
    key: 'getName',
    value: function getName() {
      return this._name;
    }
  }, {
    key: 'getFullName',
    value: function getFullName() {
      return this._type ? this._type + ': ' + this._name : this.getName();
    }
  }, {
    key: 'setType',
    value: function setType(type) {
      this._type = type;
    }
  }, {
    key: 'getType',
    value: function getType() {
      return this._type;
    }
  }, {
    key: 'setRole',
    value: function setRole(role) {
      this._role = role;
    }
  }, {
    key: 'getRole',
    value: function getRole() {
      return this._role;
    }
  }]);

  return Identity;
}();

exports.default = Identity;
module.exports = exports['default'];
