// generic-game/dist/classes/character/Character.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _events = require('events');

var _events2 = _interopRequireDefault(_events);

var _Inventory = require('./inventory/Inventory');

var _Inventory2 = _interopRequireDefault(_Inventory);

var _Equipment = require('./equipment/Equipment');

var _Equipment2 = _interopRequireDefault(_Equipment);

var _Characteristics = require('./characteristics/Characteristics');

var _Characteristics2 = _interopRequireDefault(_Characteristics);

var _Battle = require('./battle/Battle');

var _Battle2 = _interopRequireDefault(_Battle);

var _Status = require('./status/Status');

var _Status2 = _interopRequireDefault(_Status);

var _Bank = require('./bank/Bank');

var _Bank2 = _interopRequireDefault(_Bank);

var _Identity = require('./identity/Identity');

var _Identity2 = _interopRequireDefault(_Identity);

var _Experience = require('./experience/Experience');

var _Experience2 = _interopRequireDefault(_Experience);

var _Quests = require('./quests/Quests');

var _Quests2 = _interopRequireDefault(_Quests);

var _Role = require('./role/Role');

var _Role2 = _interopRequireDefault(_Role);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var Character = function () {
  function Character(_ref) {
    var _ref$identity = _ref.identity,
        identity = _ref$identity === undefined ? {} : _ref$identity,
        _ref$experience = _ref.experience,
        experience = _ref$experience === undefined ? 0 : _ref$experience,
        _ref$role = _ref.role,
        role = _ref$role === undefined ? {} : _ref$role,
        _ref$items = _ref.items,
        items = _ref$items === undefined ? [] : _ref$items,
        _ref$equipment = _ref.equipment,
        equipment = _ref$equipment === undefined ? {} : _ref$equipment,
        _ref$bank = _ref.bank,
        bank = _ref$bank === undefined ? {} : _ref$bank,
        _ref$characteristics = _ref.characteristics,
        characteristics = _ref$characteristics === undefined ? [] : _ref$characteristics,
        _ref$inventory = _ref.inventory,
        inventory = _ref$inventory === undefined ? {} : _ref$inventory;

    _classCallCheck(this, Character);

    this.identity = new _Identity2.default(identity, role);
    this.events = new _events2.default();
    this.inventory = new _Inventory2.default(inventory);
    this.role = new _Role2.default(role);
    this.equipment = new _Equipment2.default(equipment);
    this.characteristics = new _Characteristics2.default({ characteristics: characteristics });
    this.battle = new _Battle2.default({ character: this });
    this.status = new _Status2.default({ character: this });
    this.bank = new _Bank2.default(bank);
    this.experience = new _Experience2.default({ value: experience });
    this.quests = new _Quests2.default();

    this.identity.setRole(this.role.getName());
  }

  _createClass(Character, [{
    key: 'interact',
    value: function interact(thing) {
      return thing.interaction(this);
    }
  }, {
    key: 'getLevel',
    value: function getLevel() {
      return this.experience.computeLevel();
    }
  }]);

  return Character;
}();

exports.default = Character;
module.exports = exports['default'];
